### PR TITLE
fix(cdk/testing): account for preventDefault in keyboard events

### LINKED ISSUE #27475: CDK/testing: typeInElement ignores event.defaultPrevented
### Is this a regression?

- [ ] Yes, this behavior used to work in the previous version

### The previous version in which this bug was not present was

_No response_

### Description

The `typeInElement` [1] still sets the input value even if the `keydown` event has default prevented. This is different from the actual behavior [2]

[1]: https://github.com/angular/components/blob/ffffc82084b3528a4b5a95a91ec77359449acc97/src/cdk/testing/testbed/fake-events/type-in-element.ts#L114
[2]: https://stackblitz.com/edit/web-platform-gwblaq?file=script.js

### Reproduction

StackBlitz link: https://stackblitz.com/edit/harness-issue-4xv58x?file=src%2Fapp%2Fharness-example.html

### Expected Behavior

`event.preventDefault()` should stop the browser from populating the input element.

### Actual Behavior

Input element is still populated with values

### Environment

- Angular: 10.2.3
- CDK/Material: 11.0.0
- Browser(s): Chrome
- Operating System (e.g. Windows, macOS, Ubuntu): macOS

### PR DESCRIPTION
**Pull request by crisbeto**
Currently we try to mimic the user typing in the `typeInElement` utility by incrementally setting the value and dispatching the same sequence of events. The problem is that we weren't accounting for `preventDefault` which can block some keys from being assigned and some events from firing. This leads to inconsistencies between tests and the sequence of events triggered by a user. It is especially noticeable in components like the chip input where some keys act as separators.

These changes update the logic to take `preventDefault` into account and try to mimic the native event sequence as closely as possible.

Fixes #27475.

### CODE REVIEW

File: src/cdk/testing/testbed/fake-events/type-in-element.ts [hunk 1]
```diff
@@ -25,6 +35,15 @@ const incrementalInputTypes = new Set([
 /** Characters whose key code doesn't match their character code. */
 const KEYCODE_MISMATCHES: Record<string, number> = {
   ',': COMMA,
+  '.': PERIOD,
```
Comment (crisbeto) [inline src/cdk/testing/testbed/fake-events/type-in-element.ts:38]: Note that I listed just the most common conflicting characters that may show up in a test string. Our approach for mapping characters to keycodes using `charCodeAt(0)` is fundamentally flawed for many more edge cases. It has been this way since we introduced test harnesses.

Comment (andrewseguin) [reply src/cdk/testing/testbed/fake-events/type-in-element.ts:38]: Is there a better design we ought to use?

**Review by andrewseguin — APPROVED** [review]

**Comment by crisbeto** [comment]
Closing since the TGP revealed some fundamental issues with this approach and how we deal with key events in harnesses in general. We should revisit it once we've moved away from using `KeyboardEvent.keyCode`.
